deeplake/core/tensor.py

```diff
@@ -1485,20 +1485,16 @@ def delete_vdb_index(self, id: str):
 
     def _regenerate_vdb_indexes(self):
         try:
-            is_embedding = self.htype == "embedding"
-            has_vdb_indexes = hasattr(self.meta, "vdb_indexes")
-            try:
-                vdb_index_ids_present = len(self.meta.vdb_indexes) > 0
-            except AttributeError:
-                vdb_index_ids_present = False
-
-            if is_embedding and has_vdb_indexes and vdb_index_ids_present:
-                for vdb_index in self.meta.vdb_indexes:
-                    id = vdb_index["id"]
-                    distance = vdb_index["distance"]
-                    self.delete_vdb_index(id)
-                    # Recreate it back.
-                    self.create_vdb_index(id, distance)
+            vdb_indexes = self.get_vdb_indexes()
+
+            for vdb_index in vdb_indexes:
+                self.delete_vdb_index(vdb_index["id"])
+                # Recreate it back.
+                self.create_vdb_index(
+                    vdb_index["id"],
+                    vdb_index["distance"],
+                    additional_params=vdb_index.get("additional_params", None),
+                )
```

Comment:
What happens when either of these commands fail? do we end up with corrupt indexes or do we just delete index? If possible I think we should rolling back to previous index state.

Reply:
@adolkhan I don't think we should roll back, because the previous index won't have all the samples, so even though it could technically, it's not a correct index. Btw this stuff is being removed with incremental updates, so this code will likely exist for a short time.

Reply:
In the future shouldn't we undo the whole operation. Say if we tried to delete some samples, index part errored out, shouldn't we add back all deleted elements and use old index while throwing exception that this operation was unsuccessful?

Reply:
otherwise the dataset state will be in weir situation. Some samples got removed while index was wiped out completely or if error happened during deletion, then we either end up with corrupted index or old index. Isn't it the case? @sounakr correct me if I am wrong. But since you said that this code is short lived, then I think it is fine to merge these changes

Reply:
That's a good point, though I don't like the idea of writing rollbacks near the user-facing API. We should write all of this at the low level, i.e. the API for adding samples and updating the index, and all the rollback logic, should be implemented at a low level behind a single user-facing API call.

That's how `ds.extend({tensor_1.... tensor_2....}) `works. In that API, and if some samples fail to append, no samples append at all, presumably using something similar to rollbacks, but most importantly, the person who calls the `ds.extend` API doesn't implement or worry about that.

Reply:
@istranic / @adolkhan , When there is an error or corruption we can do two things.
1. Retry index creation from very beginning. i.e. not incremental but from scratch. 
2. Even if the retry fails then mark the index as non_usable. This non_usable flag will prohibit us from using the index temporairy during search options. 
3. The index has to be manually dropped and recreated later to clear out the flag and make it reusable again. 

Let me know you thoughts. Will implement this logic in incremental.

Reply:
Dataset and index should be kept separate. Dataset is a superset and index can be recreated anytime if Dataset is present. So in case of corruption we have Indra generic consine similarity and L2 implementation which can take over even if hnsw index is absent.

Reply:
@sounakr My concern is not about the index corruption itself. My concern is about our code structure, and the fact that it's architected in a way that's very bad for a developer or a user.